src/meta/processors/admin/ListClusterInfoProcessor.cpp

```diff
@@ -0,0 +1,78 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "meta/processors/admin/ListClusterInfoProcessor.h"
+
+namespace nebula {
+namespace meta {
+
+void ListClusterInfoProcessor::process(const cpp2::ListClusterInfoReq& req) {
+    UNUSED(req);
+    auto* store = dynamic_cast<kvstore::NebulaStore*>(kvstore_);
+    if (store == nullptr) {
+        onFinished();
+        return;
+    }
+
+    if (!store->isLeader(kDefaultSpaceId, kDefaultPartId)) {
+        handleErrorCode(cpp2::ErrorCode::E_LEADER_CHANGED);
+        onFinished();
+        return;
+    }
+
+    const auto& prefix = MetaServiceUtils::hostPrefix();
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto ret = kvstore_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "prefix failed:" << ret;
+        handleErrorCode(cpp2::ErrorCode::E_LIST_CLUSTER_FAILURE);
+        onFinished();
+        return;
+    }
+
+    std::vector<nebula::cpp2::NodeInfo> storages;
+    while (iter->valid()) {
+        auto host = MetaServiceUtils::parseHostKey(iter->key());
+        HostInfo info = HostInfo::decode(iter->val());
+
+        if (info.role_ != cpp2::HostRole::STORAGE) {
+            iter->next();
+            continue;
+        }
+
+        auto status = client_->listClusterInfo(host).get();
+        if (!status.ok()) {
+            LOG(ERROR) << "listcluster info from storage failed, host: " << host;
+            handleErrorCode(cpp2::ErrorCode::E_LIST_CLUSTER_FAILURE);
+            onFinished();
+            return;
+        }
+
+        storages.emplace_back(apache::thrift::FragileConstructor(),
+                              std::move(host),
+                              status.value().first,
+                              status.value().second);
+        iter->next();
+    }
+
+    auto* pm = store->partManager();
+    auto* mpm = dynamic_cast<nebula::kvstore::MemPartManager*>(pm);
+    if (mpm == nullptr) {
+        resp_.set_code(cpp2::ErrorCode::E_LIST_CLUSTER_FAILURE);
+        onFinished();
+        return;
+    }
+    auto& map = mpm->partsMap();
+    auto hosts = map[nebula::meta::kDefaultSpaceId][nebula::meta::kDefaultPartId].hosts_;
```

Comment:
Why not just merge `ListClusterInfo` and `GetMetaDirInfo` into one? Since you collect all info of storage here, but meta is handled in both processor.

Reply:
Do you mean that the interface GetMetaDirInfo is called directly in the ListClusterInfo interface?

Reply:
Because I see meta data path is returned in `GetMetaDirInfo`, but storage data path and meta hosts is returned here in `ListClusterInfo`. Looks a bit weird, why not `GetMetaDirInfo` return all meta infos, and rename this as `ListStorageInfo` or something else. Or just only use one method, return everything.

Reply:
There are two reasons for this
1 because for the time being we don't have an encapsulated interface to control another meta from a meta
2 ListClusterInfo only needs to be sent to the meta leader, while GetMetaDirInfo needs to be sent to all meta servers.